lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -490,99 +508,118 @@ isOurTx tx u
         pure . not . UTxO.null $
         u `UTxO.restrictedBy` Set.fromList (fst <$> tx ^. #resolvedInputs)
     txHasRelevantOutput =
-        F.or <$> sequence (isOurAddress . (^. #address) <$> tx ^. #outputs)
+        F.or <$> sequence (isOursState . (^. #address) <$> tx ^. #outputs)
     txHasRelevantWithdrawal =
-        F.or <$> sequence (isOurWithdrawal <$> Map.toList (tx ^. #withdrawals))
+        F.or <$> sequence (isOursState . fst <$> Map.toList (tx ^. #withdrawals))
+    
+isOursState :: IsOurs s entity => entity -> State s Bool
+isOursState x = isJust <$> state (isOurs x)
 
 {-------------------------------------------------------------------------------
-                               Internals
+    Modification of UTxO
 -------------------------------------------------------------------------------}
 
--- | Prefiltering returns all transactions of interest for the wallet. A
--- transaction is a matter of interest for the wallet if:
---
---    - It has known input(s)
---    - and/or It has known output(s)
---
--- In practice, most transactions that are of interest have an output to the
--- wallet but some may actually have no change output whatsoever and be only
+-- | Apply a 'Block' to the 'UTxO'.
+--
+-- Here, we assume that address discovery has already been performed and
+-- that the address discovery state @s@ identifies all our output addresses
+-- in the given 'Block'.
+--
+-- A 'FilteredBlock' is returned in addition to the new 'UTxO' set.
+-- This 'FilteredBlock' includes those transactions and delegations
+-- that are in the given 'Block' and that are also relevant to the wallet,
+-- i.e. they have
+-- 
+-- * Outputs with known addresses
+-- * Inputs referrring to known outputs.
+-- 
+-- In practice, most transactions that are relevant have a change output to the
```

Comment:
is it true? If the transaction has our inputs then indeed we have change also, but what if we are recepient wallet, then change is not ours almost always, right? Exception would be transactions to our own wallet

Reply:
I think that in the sentence, "change" is intended to refer to any output that goes to our wallet, not just the "proper" change outputs (with /1 derivation). I will reword this for clarity.